server/src/main/java/org/elasticsearch/index/codec/vectors/DefaultIVFVectorsReader.java

```diff
@@ -444,18 +440,36 @@ private static int collect(int[] docIds, int offset, KnnCollector knnCollector,
 
         @Override
         public int visit(KnnCollector knnCollector) throws IOException {
+            byte postingListType = indexInput.readByte();
```

Comment:
What if we just always put the block of docs at the start of every block?

So every block is
`[blk0, blk1, blk2, ... tail]
`[[encoded doc ids, vectors],...[tail encoded doc ids, vectors]]` 

We know the block size (16), we know the previous base block (if we want to delta encode eventually). 

If we ever split soar and regular docs, we can delta encode with the "doc base" (just like regular postings list).

Are we concerned about speed or just size increase?

Reply:
This is fair, I can have a go to it.

Reply:
Introduced this approach in [71e30a1](https://github.com/elastic/elasticsearch/pull/132498/commits/71e30a1faf520f2c887a7d7db0aaaf11a4171630). Much simpler.

Reply:
Checking the posting list size of 1m vectors with 1024 dims:

main: `302.780.965 bytes (318,8 MB on disk)`
PR:  `301.815.585 bytes (302 MB on disk)`

Reply:
Ah, it breaks alignment, which is frustrating.

I am testing with sorted doc-ids right now (now that we aren't skipping duplicate vectors). 

GroupVarInt also has a "single byte read" to determine the output flag. Having a single byte read for every group of 16 integers does seem weird. 

I wonder if we can do something more clever by delta-encoding all the vectors (we read all the blocks in order anyways, so we can keep the running sum), and pick the appropriate encoding that works for all the blocks. Then we can write that encoding byte at the front of the entire list, and have uniform encoding for every block. 

This might be slightly less disk efficient, but it will likely align better.